### PR TITLE
remoting: Handle DLLImport failure of libpsrpclient

### PR DESCRIPTION
**Pull request by dantraMSFT**
Issue https://github.com/PowerShell/PowerShell/issues/4029 exposed two problems when failing to load libpsrpclient. 
* WSManAPIDataCommon.ctor was not identifying the binary that failed DLLImport; hindering diagnosability 
* PrioritySendDataCollection.Clear() would throw a NullReferenceException when called from the finalizer after an error path.

This change addresses both of the above:

* Guard against null arrays in PrioritySendDataCollection.Clear() - called from finalizer in error paths.

* Diagnosability: Log the DllNotFoundException in WSManAPIDataCommon ctor and also include it as the internal exception when throwing PSRemotingTransportException.

### CODE REVIEW

**Comment by dantraMSFT** [comment]
@mirichmo and @PaulHigin will either of you be able to look at this?

File: src/System.Management.Automation/engine/remoting/fanin/PriorityCollection.cs
Comment (anmenaga) [inline src/System.Management.Automation/engine/remoting/fanin/PriorityCollection.cs]: Is it possible to lock on '_syncObjects' array itself without creating additional field "_syncObject" ?

File: src/System.Management.Automation/engine/remoting/fanin/PriorityCollection.cs
Comment (anmenaga) [inline src/System.Management.Automation/engine/remoting/fanin/PriorityCollection.cs]: This field is different from array "_syncObjects" only by last letter - it is confusing.
It would be good to either add comment for "_syncObjects" array about what it is; or rename "_syncObject"; or both.

**Review by anmenaga — COMMENTED** [review]
Leave a comment

File: src/System.Management.Automation/engine/remoting/fanin/PriorityCollection.cs
Comment (dantraMSFT) [reply src/System.Management.Automation/engine/remoting/fanin/PriorityCollection.cs]: @anmenaga: Actually, after reviewing this logic again, _syncObject is insufficient since it only guards Clear(). Additionally, _syncObjects also appears to be fragile; it guards against individual usage of a given element but assumes the elements are valid; which is the actual problem.

My thinking is _syncObjects should be removed and _syncObject used to serialize access to the set instead of individual members.

Thoughts?

**Comment by TravisEz13** [comment]
also related to issue #5566

File: src/System.Management.Automation/engine/remoting/fanin/PriorityCollection.cs
Comment (anmenaga) [reply src/System.Management.Automation/engine/remoting/fanin/PriorityCollection.cs]: @dantraMSFT I'm all for simplification, but there should be a reason why they decided to go more complex 'individual-sync-object' route in the first place... Maybe something will deadlock with a single syncObject, and parallel "Add"s won't be able to execute for different DataPriorityTypes (not sure how critical this really is).

**Comment by TravisEz13** [comment]
LGTM, but I would like @PaulHigin  to have a look.

**Comment by TravisEz13** [comment]
@anmenaga Do you want to sign off, or did you just comment?

**Comment by anmenaga** [comment]
@TravisEz13 
The conversation continues about refactoring that Dan proposed; but I think that refactoring (if needs to be done at all) can be done as separate PR.

**Review by anmenaga — APPROVED** [review]
LGTM

**Comment by dantraMSFT** [comment]
@TravisEz13 This should be ready to merge.

**Comment by TravisEz13** [comment]
I've been asked to wait for a review by @PaulHigin  or @daxian-dbw

File: src/System.Management.Automation/engine/remoting/fanin/PriorityCollection.cs
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/remoting/fanin/PriorityCollection.cs]: This assert should be removed because apparently `_dataToBeSent` can be null in that error path.

File: src/System.Management.Automation/engine/remoting/fanin/PriorityCollection.cs
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/remoting/fanin/PriorityCollection.cs]: I think we can avoid the sync object `_syncObject` by the following code:
```CSharp
if (null != _dataSyncObjects)
{
    int promptResponseIndex = (int)DataPriorityType.PromptResponse;
    int defaultIndex = (int)DataPriorityType.Default;

    if (_dataToBeSent[promptResponseIndex] != null)
    {
        lock (_dataSyncObjects[promptResponseIndex])
        {
            if (_dataToBeSent[promptResponseIndex] != null)
            {
                _dataToBeSent[promptResponseIndex].Dispose();
                _dataToBeSent[promptResponseIndex] = null;
            }
        }
    }

    if (_dataToBeSent[defaultIndex] != null)
    {
        lock (_dataSyncObjects[defaultIndex])
        {
            if (_dataToBeSent[defaultIndex] != null)
            {
                _dataToBeSent[defaultIndex].Dispose();
                _dataToBeSent[defaultIndex] = null;
            }
        }
    }
}
```

**Review by daxian-dbw — DISMISSED** [review]
Left two comments.

File: src/System.Management.Automation/engine/remoting/fanin/PriorityCollection.cs
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/remoting/fanin/PriorityCollection.cs]: > Actually, after reviewing this logic again, _syncObject is insufficient since it only guards Clear().

`_dataSyncObjects` and `_dataToBeSent` are used in multiple methods, but it appears that there is an assumption that those methods won't be called until the `Fragmentor` property is set (where both `_dataSyncObjects` and `_dataToBeSent` would be initialized). With this assumption, I think it's fine to directly use `_dataSyncObjects` and `_dataToBeSent` in methods other than `Clear()`. But we probably should add a comment about that assumption.

For `Clear()`, it's possible to be called before `Fragmentor` being set (in error path), so it's necessary to check `_dataSyncObjects` for null beforehand.

File: src/System.Management.Automation/engine/remoting/fanin/PriorityCollection.cs
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/remoting/fanin/PriorityCollection.cs]: Never mind this comment. It's certainly right to assume `_dataToBeSent != null` when `null != _dataSyncObjects` 😄

File: src/System.Management.Automation/engine/remoting/fanin/PriorityCollection.cs
Comment (dantraMSFT) [reply src/System.Management.Automation/engine/remoting/fanin/PriorityCollection.cs]: You're right; I'll make the change.

File: src/System.Management.Automation/engine/remoting/fanin/PriorityCollection.cs
Comment (dantraMSFT) [reply src/System.Management.Automation/engine/remoting/fanin/PriorityCollection.cs]: Will do.

File: src/System.Management.Automation/engine/remoting/fanin/PriorityCollection.cs
Comment (dantraMSFT) [reply src/System.Management.Automation/engine/remoting/fanin/PriorityCollection.cs]: fixed

File: src/System.Management.Automation/engine/remoting/fanin/PriorityCollection.cs
Comment (dantraMSFT) [reply src/System.Management.Automation/engine/remoting/fanin/PriorityCollection.cs]: fixed

**Comment by dantraMSFT** [comment]
@daxian-dbw do you have any additional feedback?

**Comment by dantraMSFT** [comment]
@PaulHigin, @daxian-dbw, please take a look at this; it's marked as GA. Thanks.

**Review by daxian-dbw — APPROVED** [review]

File: src/System.Management.Automation/engine/remoting/fanin/PriorityCollection.cs
Comment (PaulHigin) [inline src/System.Management.Automation/engine/remoting/fanin/PriorityCollection.cs]: Sorry for the late review.  I am not familiar with the error path but if _dataToBeSent array can contain null entries then it seems like so can _dataSyncObjects (_dataSyncObjects[promptResponseIndex] == null ?).

It seems like these arrays should be created in the PrioritySendDataCollection constructor...

Comment (dantraMSFT) [reply src/System.Management.Automation/engine/remoting/fanin/PriorityCollection.cs]: @PaulHigin: _dataSyncObjects is populated when it is created.  The entries are not null if the array is not null.
As far as when it is created, it appears the intent was to create both arrays at the same point.  I saw no strong reason to change it.

Comment (PaulHigin) [reply src/System.Management.Automation/engine/remoting/fanin/PriorityCollection.cs]: But the _dataSyncObjects array is filled within a loop:
```
for (int i = 0; i < names.Length; i++)
    {
        _dataToBeSent[i] = new SerializedDataStream(_fragmentor.FragmentSize);
        _dataSyncObjects[i] = new object();
    }
```
If an exception is thrown within the loop then it is possible an array element is null.  Is this not the error path we are worried about?

Comment (dantraMSFT) [reply src/System.Management.Automation/engine/remoting/fanin/PriorityCollection.cs]: SerializedDataStream's ctor does not currently throw.

Note that it should (it relies on Dbg.Assert) and I opened an issue for this: https://github.com/PowerShell/PowerShell/issues/5800

**Pull request merged**
